attestation: add high-level introduction to attestation (#179)
confidential-containers/confidentialcontainers.org: content/en/docs/attestation/background/_index.md

---
title: What is Attestation?

description: Background on attestation 
weight: 1
categories:
- attestation
tags:
- trustee
- attestation
---

## Overview

Generally speaking, confidential computing has two dimensions.

> First, there are runtime isolation technologies such as memory encryption.

Reviewer: Avoid burying the lead :-D ..
```suggestion
First, there are runtime isolation technologies such as memory encryption.
Second, there is remote attestation, which enables remotely assessing those security properties before trusting it.
```

The hardware root of trust attests to the configuration and initial state of the TEE,
typically by signing a report about the TEE with a key linked to the hardware manufacturer.

Attestation goes beyond primitives such as measurement and report signing
to encompass a process for establishing trust between parties. 
Often, attestation is described from the perspective of the TEE,
but this can lead to confusion.

In its simplest form, attestation involves two parties; the hardware manufacturer
and the owner of a workload that will be deployed in a TEE.
These parties already trust each other.
The goal is for the workload owner to trust a TEE environment setup by a mostly untrusted host.
Most people describe this as a guest attesting to a workload owner or a relying party.
This is a reasonable simplification where the hardware evidence represents
the hardware manufacturer more generally.

Trustee conceptualizes this process in terms of resource requests.
When Trustee receives a request for a resource, Trustee determines whether
the requester is trustworthy before granting access to the resource.

## Disambiguation

Unfortunately, the term attestation is slightly overloaded.

> Attestation can also refer to host attestation, which is often based on TPMs.

Reviewer: Might be clearer to list the variations up front and then explain each succinctly. 
```suggestion
Besides Confidential Computing (CC) Attestation, it can also refer to 
* TPM-based host platform attestation.
* Software-based SPIFFE node or workload attestation.
```

While confidential attestation can involve vTPMs,

> confidential attestation is scoped to a TEE

Reviewer: I haven't seen the term confidential attestation before.
Consider introducing "Confidential Computing (CC) Attestation" above and then the subsequent references in disambiguation can be CC Attestation.
Since this is CoCo we don't need to be fully general about CC attestation and we can be specific to the kata utility guest vm.
```suggestion
CC attestation is typically scoped to a guest VM, here the CoCo VM housing the pod.
While CC attestation can involve vTPMs,
those vTPMs are implemented as part of the guest VM.
```

> and usually ignores the untrusted host.

Reviewer: ```suggestion
and is designed not to rely on untrusted aspects of the host.
```

Host attestation usually has a continuous and non-blocking model 
where hosts are periodically attested.
If attestation fails, some corrective action, such as quarantining the node,

> can be taken.

Reviewer: ```suggestion
Host Platform Attestation (TPM-based): Measures and verifies the integrity of the bare-metal host hardware, firmware (UEFI/BIOS), and early boot components (bootloader, host OS kernel). It is typically used to prove to the infrastructure owner that the underlying physical server is running authentic, untampered system software.
If attestation fails, some corrective action, such as quarantining the node,
can be taken.
```

> Confidential attestation, on the other hand, usually blocks the delivery of secrets

Reviewer: ```suggestion
CC attestation, on the other hand, usually blocks the delivery of secrets
```

to a TEE, and it is often scoped to the initial state of a guest.

There is overlap between these two approaches,
but crucially the party that can validate a host environment,
and the party that can validate a guest environment
are usually not the same. 

> Thus, there are two distinct processes.

Reviewer: ```suggestion
CC Attestation: As used by CoCo, proves that the Pod is protected from the infrastructure provider. It is in service of the workload owner.
CC attestation usually gates the delivery of secrets.
If attestation verification fails the Pod won't get its secrets.

```

SPIFFE/SPIRE also uses the term attestation.
With SPIRE, node attestation is somewhat similar to host attestation,
but workload attestation refers more generally to one component

> attesting to the properties of another.

Reviewer: ```suggestion
SPIFFE/SPIRE also uses the term attestation for Node and Workload use cases.
Node attestation is somewhat similar to host platform attestation.
Workload attestation relies on a trusted agent assessing some property of the workload which may not have cryptographic utility in the way that CC memory measurement does.
Both of these SPIFFE/SPIRE use cases are in service of the K8s admin.
If attestation fails, the control plane will refuse the node or the workload respectively.
```

## Workload Design

Trustee and the KBS Protocol do not specify how or when a workload
should invoke the attestation process.
Of course, attestation must take place to get access to a resource from Trustee.
With Confidential Containers, attestation happens lazily; only when the first
operation that depends on a resource takes place.
This could happen prior to container startup to get access to an image decryption key
or a sealed secret, or it could happen later.
In Confidential Containers, if a workload does not depend on any resources, it won't be attested.
For various reasons, this would not be a secure workload.
On the other hand, if a workload does depend on a resource, and it receives that resource
it is implied that attestation has taken place, particularly if the resource

> is a secret known only to Trustee.

Reviewer: ```suggestion
To make full use of CC, workloads should be designed to rely on resources provisioned through Trustee.
It is only through that control that attestation is triggered and the workload's security is assessed.
```
